config/dynamic.c

```diff
@@ -16,11 +16,536 @@
+			"dyn_cfg: %s!\n", lua_tostring(lua_state, -1));
+		ret = -1;
+		goto out;
+	}
+
+	return 0;
+
+out:
+
+	return ret;
+}
+
+static void
+handle_client(int server_socket_fd, struct dynamic_config *dy_conf)
+{
+	int ret;
+	int conn_fd;
+	socklen_t len;
+	int rcv_buff_size;
+	struct sockaddr_un client_addr;
+
+	/* The lua state used to handle the dynamic configuration files. */
+	lua_State *lua_state;
+
+	len = sizeof(client_addr);
+	conn_fd = accept(server_socket_fd,
+		(struct sockaddr *)&client_addr, &len);
+	if (conn_fd < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to accept a new connection - (%s)\n",
+			strerror(errno));
+		return;
+	}
+
+	if (unlikely(client_addr.sun_family != AF_UNIX)) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"Unexpected condition: unknown client type %d at %s\n",
+			client_addr.sun_family, __func__);
+		goto close_fd;
+	}
+
+	/*
+	 * XXX The request must be received under a specified timeout,
+	 * or the request is aborted. This code needs further test.
+	 */
+	ret = setsockopt(conn_fd, SOL_SOCKET, SO_RCVTIMEO,
+		(const char*)&dy_conf->rcv_time_out, sizeof(struct timeval));
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to call setsockopt(SO_RCVTIMEO) - (%s)\n",
+			strerror(errno));
+		goto close_fd;
+	}
+
+	rcv_buff_size = MSG_MAX_LEN;
+	ret = setsockopt(conn_fd, SOL_SOCKET,
+		SO_RCVBUF, &rcv_buff_size, sizeof(rcv_buff_size));
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to call setsockopt(SO_RCVBUF) with size = %d - (%s)\n",
+			rcv_buff_size, strerror(errno));
+		goto close_fd;
+	}
+
+	lua_state = luaL_newstate();
+	if (lua_state == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: failed to create new Lua state!\n");
+		goto close_fd;
+	}
+
+	/* Set up the Lua state while there is a connection. */
+	ret = setup_dy_lua(lua_state);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to set up the lua state\n");
+		goto close_lua;
+	}
+
+	while (1) {
+		ret = process_single_cmd(conn_fd, lua_state);
+		if (ret != 0)
+			break;
+	}
+
+close_lua:
+	lua_close(lua_state);
+
+close_fd:
+	ret = close(conn_fd);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to close the connection socket - (%s)\n",
+			strerror(errno));
+	}
+}
+
+static int
+dyn_cfg_proc(void *arg)
+{
+	int ret = 0;
+	uint32_t lcore = rte_lcore_id();
+	struct dynamic_config *dy_conf = arg;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"dyn_cfg: the Dynamic Config block is running at lcore = %u\n", lcore);
+
+	while (likely(!exiting)) {
+		fd_set fds;
+		struct timeval stv;
+ 
+		FD_ZERO(&fds);
+		FD_SET(dy_conf->sock_fd, &fds);
+ 
+		/*
+		 * 10000 usecs' timeout for the select() function.
+		 * This parameter can prevent the select() function
+		 * from blocking forever. So, the whole program can
+		 * exit when receiving a quitting signal.
+		 */
+		stv.tv_sec = 0;
+		stv.tv_usec = 10000;
+
+		ret = select(dy_conf->sock_fd + 1, &fds, NULL, NULL, &stv);
+		if (ret < 0 && errno != EINTR) {
```

Comment:
Notice that for the case I'm pointing out `ret == -1` and `errno == EINTR` and `exiting == false`, so it won't hit the first, second, or third `if` statement. In this case, the body of the loop, which expects `ret > 0`, will run.

My suggestion on how to rewrite the `if` statements and the body of the loop:
```
if (ret < 0 && errno != EINTR) {
	/* The same as it is now in the code. */
} else if (likely(ret <= 0)
	continue;

/*
 * The config component accepts only one connection at a time.
 */
RTE_VERIFY(FD_ISSET(dy_conf->sock_fd, &fds));
handle_client(dy_conf->sock_fd, dy_conf);
```

I get away without testing `exiting` because it is in `while` test, which will be tested after the `continue`. I don't need to test for `errno == EINTR` in the second `if` statement because if `ret < 0` there, `errno == EINTR` must be true. Finally, the body of the loop has a guarantee that we have reached it under the expected condition.